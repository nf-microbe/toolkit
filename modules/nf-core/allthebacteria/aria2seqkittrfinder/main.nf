process ALLTHEBACTERIA_ARIA2SEQKITTRFINDER {
    tag "${meta.id}"
    label 'process_high'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/seqkit_aria2_biopython:2bb0d76d26291e6d' :
        'community.wave.seqera.io/library/seqkit_aria2_biopython:2256c902d3a46191' }"

    input:
    tuple val(meta), val(url)

    output:
    tuple val(meta), path("${prefix}.trfinder.fasta.gz")    , emit: fasta
    tuple val(meta), path("${prefix}.trfinder.tsv")         , emit: tsv
    path "versions.yml"                                     , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def url_list    = url.collect { urls -> urls.toString() }.join(',')
    def args        = task.ext.args ?: ''
    def args2       = task.ext.args2 ?: ''
    def args3       = task.ext.args3 ?: ''
    prefix          = task.ext.prefix ?: "${meta.id}"
    """
    mkdir -p tmp/download tmp/seqkit tmp/trfinder
    IFS=',' read -r -a url_array <<< "${url_list}"
    printf '%s\\n' "\${url_array[@]}" > aria2_file.tsv

    ### Download AllTheBacteria assemblies
    aria2c \\
        --input-file=aria2_file.tsv \\
        --dir=tmp/download/ \\
        --max-connection-per-server=${task.cpus} \\
        --split=${task.cpus} \\
        --max-concurrent-downloads=${task.cpus} \\
        ${args}

    tar -xvf tmp/download/*.tar.xz -C tmp/download
    rm tmp/download/*.tar.xz

    ### Remove short contigs
    for dir in tmp/download/*; do
        for file in \$dir/*; do
            filename=\$(basename \$file)

            seqkit \\
                seq \\
                --threads ${task.cpus} \\
                ${args2} \\
                \$file \\
                --out-file tmp/seqkit/\${filename%.*}.fasta
        done
    done

    rm -rf tmp/download/

    ### Identify Terminal repeats
    cd tmp/trfinder

    for file in ../seqkit/*.fasta; do
        filename=\$(basename \$file)

        trfinder.py \\
            --input \$file \\
            --prefix ATB_\${filename%.*} \\
            ${args3}
    done

    cd ../..
    rm -rf tmp/seqkit/

    # combine and compress tr finder files
    cat tmp/trfinder/*.trfinder.fasta > ${prefix}.trfinder.fasta
    awk '(NR == 1) || (FNR > 1)' tmp/trfinder/*.trfinder.tsv > ${prefix}.trfinder.tsv

    rm -rf tmp/trfinder/
    gzip -f ${prefix}.trfinder.fasta

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        aria2: \$(echo \$(aria2c --version 2>&1) | grep 'aria2 version' | cut -f3 -d ' ')
        seqkit: \$(seqkit version | cut -d' ' -f2)
        python: \$( python --version | sed 's/Python //' )
        biopython: \$(python -c "import Bio; print(Bio.__version__)")
    END_VERSIONS
    """

    stub:
    def args   = task.ext.args ?: ''
    def args2   = task.ext.args2 ?: ''
    def args3   = task.ext.args3 ?: ''
    prefix  = task.ext.prefix ?: "${meta.id}"
    """
    echo "" | gzip > ${prefix}.trfinder.fasta.gz
    touch ${prefix}.trfinder.tsv

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        aria2: \$(echo \$(aria2c --version 2>&1) | grep 'aria2 version' | cut -f3 -d ' ')
        seqkit: \$(seqkit version | cut -d' ' -f2)
        python: \$( python --version | sed 's/Python //' )
        biopython: \$(python -c "import Bio; print(Bio.__version__)")
    END_VERSIONS
    """
}
