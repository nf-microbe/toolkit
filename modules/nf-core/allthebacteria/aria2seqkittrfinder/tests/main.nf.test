nextflow_process {

    name "Test Process ALLTHEBACTERIA_ARIA2SEQKITTRFINDER"
    script "../main.nf"
    process "ALLTHEBACTERIA_ARIA2SEQKITTRFINDER"

    tag "modules"
    tag "modules_nfcore"
    tag "allthebacteria"
    tag "allthebacteria/aria2seqkittrfinder"

    test("atb.assembly.incr_release.202408.batch.104.tar.xz") {
        when {
            process {
                """
                input[0] = [
                    [ id:'atb.assembly.incr_release.202408.batch.104.tar.xz' ], // meta map
                    [
                        "https://osf.io/download/66d9b3294d20d12534a3cd88/"
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("atb.assembly.incr_release.202408.batch.104.tar.xz - stub") {
        options "-stub"
        when {
            process {
                """
                input[0] = [
                    [ id:'atb.assembly.incr_release.202408.batch.104.tar.xz' ], // meta map
                    [
                        "https://osf.io/download/66d9b3294d20d12534a3cd88/"
                    ]
                ]
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
