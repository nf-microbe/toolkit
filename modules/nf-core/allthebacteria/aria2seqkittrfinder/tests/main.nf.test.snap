{
    "atb.assembly.incr_release.202408.batch.104.tar.xz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.fasta.gz:md5,6decac710cbcbcef855ddb61f5a943f2"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.tsv:md5,abc053a04761504ede8b69c7095342d2"
                    ]
                ],
                "2": [
                    "versions.yml:md5,4d537387f97a1af2fac93bb00835f3e4"
                ],
                "fasta": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.fasta.gz:md5,6decac710cbcbcef855ddb61f5a943f2"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.tsv:md5,abc053a04761504ede8b69c7095342d2"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4d537387f97a1af2fac93bb00835f3e4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-21T09:49:30.125046494"
    },
    "atb.assembly.incr_release.202408.batch.104.tar.xz - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,4d537387f97a1af2fac93bb00835f3e4"
                ],
                "fasta": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.fasta.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "tsv": [
                    [
                        {
                            "id": "atb.assembly.incr_release.202408.batch.104.tar.xz"
                        },
                        "atb.assembly.incr_release.202408.batch.104.tar.xz.trfinder.tsv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,4d537387f97a1af2fac93bb00835f3e4"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "24.04.4"
        },
        "timestamp": "2024-10-21T09:51:25.080630255"
    }
}